algosdk/atomic_transaction_composer.py

```diff
@@ -0,0 +1,529 @@
+        for t in self.txn_list:
+            t.txn.group = group_id
+            self.tx_ids.append(t.txn.get_txid())
+
+        self.status = AtomicTransactionComposerStatus.BUILT
+        return self.txn_list
+
+    def gather_signatures(self):
+        """
+        Obtain signatures for each transaction in this group. If signatures have already been obtained,
+        this method will return cached versions of the signatures.
+        The composer's status will be at least SIGNED after executing this method.
+        An error will be thrown if signing any of the transactions fails.
+
+        Returns:
+            list[SignedTransactions]: list of signed transactions
+        """
+        if self.status >= AtomicTransactionComposerStatus.SIGNED:
+            # Return cached versions of the signatures
+            return self.signed_txns
+
+        stxn_list = [None] * len(self.txn_list)
+        signer_indexes = {}  # Map a signer to a list of indices to sign
+        txn_list = self.build_group()
+        for i, txn_with_signer in enumerate(txn_list):
+            if txn_with_signer.signer not in signer_indexes:
+                signer_indexes[txn_with_signer.signer] = []
+            signer_indexes[txn_with_signer.signer].append(i)
+
+        # Sign then merge the signed transactions in order
+        txns = [t.txn for t in self.txn_list]
+        for signer, indexes in signer_indexes.items():
+            stxns = signer.sign_transactions(txns, indexes)
+            for i, stxn in enumerate(stxns):
+                index = indexes[i]
+                stxn_list[index] = stxn
+
+        if None in stxn_list:
+            raise error.AtomicTransactionComposerError(
+                "missing signatures, got {}".format(stxn_list)
+            )
+
+        self.status = AtomicTransactionComposerStatus.SIGNED
+        self.signed_txns = stxn_list
+        return self.signed_txns
+
+    def submit(self, client):
+        """
+        Send the transaction group to the network, but don't wait for it to be
+        committed to a block. An error will be thrown if submission fails.
+        The composer's status must be SUBMITTED or lower before calling this method.
+        If submission is successful, this composer's status will update to SUBMITTED.
+
+        Note: a group can only be submitted again if it fails.
+
+        Args:
+            client (AlgodClient): Algod V2 client
+
+        Returns:
+            list[Transaction]: list of submitted transactions
+        """
+        if self.status <= AtomicTransactionComposerStatus.SUBMITTED:
+            self.gather_signatures()
+        else:
+            raise error.AtomicTransactionComposerError(
+                "AtomicTransactionComposerStatus must be submitted or lower to submit a group"
+            )
+
+        client.send_transactions(self.signed_txns)
+        self.status = AtomicTransactionComposerStatus.SUBMITTED
+        return self.tx_ids
+
+    def execute(self, client, wait_rounds):
```

Comment:
Do we want to have a default number of `wait_rounds`. I suspect the answer is "no" but it's worth your 2-cents.

Reply:
I think `wait_rounds=0` could be nice (no timeout), but I just left it as is so it could be consistent with the other SDKs.

Reply:
👍

Reply:
I think `wait_rounds=0` is actually something you should never do in practice. algod currently has a bug where it will never return an error for a failed transaction in some cases (see https://github.com/algorand/go-algorand-internal/issues/1759), meaning your code would be blocked if you were unlucky enough to encounter that bug.

Besides, `wait_rounds=1000` is the maximum you should ever need to wait, since that's a max lifetime of a transaction. This _might_ be a good candidate for a default value if we want one, but it's still pretty large.

Reply:
Interesting, I didn't know about that first point. I think `wait_for_confirmation()` in this SDK actually has a default value for `wait_rounds=0`: https://github.com/algochoi/py-algorand-sdk/blob/e444328616d99c88b830366dedcbf75e9795dcb3/algosdk/future/transaction.py#L3013